### PR TITLE
Add additional upstream USB midi port to seperate MPE data 

### PR DESCRIPTION
**Pull request by m-m-adams**
So far just broke out and commented the existing USB configuration with field names and options from the USB Midi standard

Note - lsusb shows the endpoint descriptions as invalid but they seem to work fine. Just noting it so future tests don't look like a regression.

### CODE REVIEW

**Comment by m-m-adams** [comment]
Current status - successfully presents 3 ports to host and doesn't break anything else.

Current behaviour is sends go out only on port 1, midi received on all 3 ports is summed

**Comment by m-m-adams** [comment]
This now fully implements support for 2 upstream USB ports, including input, output, and settings. The connectedUSBMIDIDevice object now contains an array of USBMIDIDevices rather than a single value. The USBMIDIDevice object is used primarily for settings and in this case more appropriately represents a USB MIDI cable than a device.

As the cable in a USB midi packet is 0 indexed, this is used directly on receive to select the appropriate port if it has been initialized. On sends, the sendUsbMidi function now iterates through all initialized ports on all usb midi devices, and if one exists and wants a midi packet the index of the USBMIDIDevice array is used directly as the cable number.

The driver configuration is modified to present as 3 USB midi interfaces to support future expansion. While only 2 can be directly used by the Deluge right now (16 midi channels +2 MPE zones) future applications such as SYSEX or DAW controller support may require an extra port.

There is a commit enabling HS usb support however it has been reverted back to FS. HS USB works, however some USB midi hubs do not operate in HS mode and additional code will need to be added to revert back to FS mode if HS is unsuccessful. 

In order to support USB midi devices with multiple ports while the Deluge is in Host mode, the device array is set to hold 4, however currently only the first device is initialized in host mode. Additional changes will be needed to the USB MIDI Host code to support multiple cables.

**Comment by alex-r-t** [comment]
Do you have a recommended test procedure to make sure everything is working the way you expect?

**Comment by m-m-adams** [comment]
> Do you have a recommended test procedure to make sure everything is working the way you expect?

1. A single midi device with 3 ports should appear on the OS side, this is tested via pytest in the tests directory. I'm not able to test on windows so confirming it works there would be great.

2. The settings menu should correctly work to configure MPE zones and bend ranges for both ports, testing 7seg is especially useful since I don't have one but also just more eyes on it in case I missed an edge case

3. Verify that normal midi and MPE zones are correctly sent to each port as per the settings. This part I'm confident works but again I might have missed an edge case 

4. Backwards compatibility on the first port - it should pick up the existing settings without issue, and any songs learned with 'differentiate inputs' on should still work, but this is the kind of thing that could really use more testing to make sure

File: src/MIDIDevice.h [hunk 1]
```diff
@@ -93,6 +101,7 @@ class MIDIDevice {
 
 	inline bool hasDefaultVelocityToLevelSet() { return defaultVelocityToLevel; }
 
+	//Only 2 ports per device, but this is functionally set in stone due to existing code
 	MIDIPort ports
 	    [2]; // I think I used an array here so the settings menu could deal with either one easily - which doesn't seem like a very strong reason really...
```
[the remark this answers is not in the document]
Comment (m-m-adams) [reply src/MIDIDevice.h:106]: I couldn't find any better reasons for making it an array so I imagine it was for convenience in the settings. I'll reformat the existing comment

File: src/MIDIDeviceManager.cpp [hunk 2]
```diff
@@ -190,7 +197,7 @@ void recountSmallestMPEZones() {
 }
 
 extern "C" void hostedDeviceConfigured(int ip, int midiDeviceNum) {
-
+	//well we're at it should recognize multiple virtual ports here too
```
[the remark this answers is not in the document]
Comment (m-m-adams) [reply src/MIDIDeviceManager.cpp:200]: Yes - I'll remove it

File: src/MIDIDeviceManager.cpp [hunk 3]
```diff
@@ -230,35 +241,48 @@ extern "C" void hostedDeviceDetached(int ip, int midiDeviceNum) {
...
+	for (int i = 0; i <= ports; i++) {
+		MIDIDeviceUSB* device = connectedDevice->device[i];
+		if (device) { // Surely always has one?
 
-	MIDIDeviceUSB* device = connectedUSBMIDIDevices[ip][midiDeviceNum].device;
-	if (device) { // Surely always has one?
+			device->connectionFlags &= ~(1 << midiDeviceNum);
 
-		device->connectionFlags &= ~(1 << midiDeviceNum);
+			recountSmallestMPEZones();
```
[the remark this answers is not in the document]
Comment (m-m-adams) [reply src/MIDIDeviceManager.cpp:252]: Will do

**Comment by bfredl** [comment]
> There is a commit enabling HS usb support however it has been reverted back to FS. HS USB works, however some USB midi hubs do not operate in HS mode and additional code will need to be added to revert back to FS mode if HS is unsuccessful.

Would it be realistic to make HS USB a run-time option in the global settings menu?

**Comment by m-m-adams** [comment]
> > There is a commit enabling HS usb support however it has been reverted back to FS. HS USB works, however some USB midi hubs do not operate in HS mode and additional code will need to be added to revert back to FS mode if HS is unsuccessful.
> 
> Would it be realistic to make HS USB a run-time option in the global settings menu? 

Yeah probably, something like the Elektron USB menu where you select between midi, stereo audio+midi, and 'delugerbridge' would make sense and the first two can use USB FS. Since it takes more than a simple change I think it belongs in a separate PR though.

File: src/MIDIDeviceManager.cpp [hunk 4]
```diff
@@ -49,18 +49,23 @@ struct {
 	uint16_t productId;
 } usbDeviceCurrentlyBeingSetUp[USB_NUM_USBIP];
 
-MIDIDeviceUSBUpstream upstreamUSBMIDIDevice;
+//This class represents a thing you can send midi too,
+//the virtual cable is an implementation detail
+MIDIDeviceUSBUpstream upstreamUSBMIDIDevice_port1;
+MIDIDeviceUSBUpstream upstreamUSBMIDIDevice_port2;
```
[the remark this answers is not in the document]
Comment (m-m-adams) [reply src/MIDIDeviceManager.cpp:55]: I'm looking at it with fresh eyes and I think it's better left as two independent variables. The deluge is only able to sequence 32 total channels (2 MPE zones + 16) so adding more user configurable ports would not be useful, and would also require changes to the USB midi config. Keeping them as independent structures better reflects the fact that it is not easily expandable to additional ports. 

These are effectively just settings structures for MPE configuration, they're both held in an array under the connectedUSBMIDIDevice that controls the actual send and receive functionality

**Comment by jamiefaye** [comment]
I am not seeing any Deluge stuff on my Mac OS MIDI Monitor. I did remove the USB cable when starting the Deluge in the debugger. If there is anything I have to do to turn this on, let me know. (It should "just work like it always has"). I also have not found anything in the manuals about a global settings menu, so if anyone can clue this gitiot in on what that means, please do.

**Comment by m-m-adams** [comment]
> 

Confirmed, I use a USB midi interface with my mac and didn't try connecting directly. Will update when I figure out a fix

**Comment by m-m-adams** [comment]
> I am not seeing any Deluge stuff on my Mac OS MIDI Monitor. I did remove the USB cable when starting the Deluge in the debugger. If there is anything I have to do to turn this on, let me know. (It should "just work like it always has"). I also have not found anything in the manuals about a global settings menu, so if anyone can clue this gitiot in on what that means, please do.

It should work now, connecting on Linux works with the same config pointer in the FS and HS parts of the Renesas driver's USB connection struct but that was causing it to fail on connection to MacOS. 


Tips on Clang format would be appreciated, my local copy and the workflow seem to disagree on the last file

**Comment by m-m-adams** [comment]
> I am not seeing any Deluge stuff on my Mac OS MIDI Monitor. I did remove the USB cable when starting the Deluge in the debugger. If there is anything I have to do to turn this on, let me know. (It should "just work like it always has"). I also have not found anything in the manuals about a global settings menu, so if anyone can clue this gitiot in on what that means, please do.

@jamiefaye the settings menu I meant was shift+select, then midi, then devices, then you should see the two upstream ports under DIN Ports. Upstream USB 1 should copy the existing settings for the upstream USB, and upstream USB2 should default to 0 MPE input and output channels

**Comment by m-m-adams** [comment]
No it should be the same - the only change is this enables MPE to be seperated from regular midi. If you set upstream port 2's MPE range to be 7 in and out for the upper and lower zones, and port 1's MPE zones to 0 in both both directions, then all your MPE devices will work on port 1 and the non-MPE devices will work on port 2. 

Currently since MPE devices are sharing 16 channels with non-MPE data you have to reduce the size of the zones to accomodate other gear. With this change you can use the full MPE ranges and still sequence additional devices

**Comment by m-m-adams** [comment]
@jamiefaye here's an example screenshot - you can see the MPE notes on channel 8, port 1 after the normal midi note on the same channel but port 2

ADB8CFAA-9411-4C8C-8E88-CE71D6ED1956


90

**Comment by jamiefaye** [comment]
Aria - if you can confirm this a working on OLED, please merge it in.

**Comment by litui** [comment]
Alas, I don't own an OLED model Deluge 😅  but I'll happily accept confirmation from someone who does have one.

**Comment by alifeinbinary** [comment]
Hi gang, I've just installed the Nightly build 1.0.0-NIGHTLY-895896A and connected my Osmose to the (OLED)Deluge's USB port to test feature 3.3 but I'm not seeing the dialog in Settings > MIDI > Devices > Osmose where a port can be specified. The Osmose sends normal MIDI from Port 1 and MPE from Port 2, however, it's aggregating the input from both ports without having access to this feature. @m-m-adams, when the Osmose is connected to a computer via USB, both Osmose Port 1 and Osmose Port 2 populate in the system MIDI devices list. Thanks for putting this feature into the firmware, hope I can help with this.

**Comment by m-m-adams** [comment]
> Hi gang, I've just installed the Nightly build 1.0.0-NIGHTLY-895896A and connected my Osmose to the (OLED)Deluge's USB port to test feature 3.3 but I'm not seeing the dialog in Settings > MIDI > Devices > Osmose where a port can be specified. The Osmose sends normal MIDI from Port 1 and MPE from Port 2, however, it's aggregating the input from both ports without having access to this feature. @m-m-adams, when the Osmose is connected to a computer via USB, both Osmose Port 1 and Osmose Port 2 populate in the system MIDI devices list. Thanks for putting this feature into the firmware, hope I can help with this. 

Opened issue 416 - the menu and aggregation should be fixable

**Comment by songsta** [comment]
Hi all. I'm running the final community build but not seeing any additional ports available in the midi settings menu. OLED Deluge.

**Comment by alifeinbinary** [comment]
@songsta It's assumed that Port 1 is standard MIDI and Port 2 is for MPE. Do you have a MPE device assigns its ports differently? If so, which one and how do you have it connected to your Deluge?

**Comment by songsta** [comment]
Excuse the delay in reply.  I have a Linnstrument and It's connected over USB. I don't think it sends via multiple ports.

**Pull request merged**